nixos/codeberg-pages: init

NotAShelf opened the pull request:
[Codeberg Pages]: https://codeberg.org/Codeberg/pages-server

Adds a NixOS module for [Codeberg Pages].

Current implementation is still a bit rough around the edges, but the module
should be ready for a review regardless. I've set this as a draft until I can
get tests to work reliably, tips welcome. I also would like to harden the
service a little, but it seems particularly picky with its permissions.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See
  [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests)
    (look inside
    [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or
    [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in
    [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or
    [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are
    [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package)
    to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using
      `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all
      changes have to be committed, also see
      [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in
      `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md)
  (or backporting
  [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md)
  and
  [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md)
  Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or
        breaking
  - [ ] (Module updates) Added a release notes entry if the change is
        significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS
        module
- [ ] Fits
      [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

NotAShelf committed:
nixos/codeberg-pages: init

ambroisie commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,148 @@
+
+  config = mkIf cfg.enable {
+    systemd = {
+      tmpfiles.rules = [ "d '${cfg.stateDir}' 0750 ${cfg.user} ${cfg.group} - -" ];
```
Please use `tmpfiles.settings` instead.

```suggestion
      tmpfiles.settings = {
        "10-codeberge-pages" = {
          ${cfg.statedir} = {
            d = {
              inherit (cfg) group user;
              mode = "0750";
            };
          };
        };
      };
```

ambroisie commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,148 @@
+
+      services.codeberg-pages = {
+        description = "Codeberg Pages Server";
+        after = [ "network.target" ];
```
IIRC correctly relying on `network.target` is discouraged (it doesn't indicate what you'd usually expect: it's not about network connectivity).

```suggestion
```

ambroisie commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,148 @@
+          Type = "simple";
+          User = cfg.user;
+          Group = cfg.group;
+          StateDirectory = cfg.stateDir;
```
`StateDirectory` will create e.g: `/var/lib/var/lib/codeberge-pages` (i.e: a subdirectory of `/var/lib`).

You probably want to use _either_ `StateDirectory` or `tmpfiles` (I'd rely on `StateDirectory` unless a use-case for storing it outside of `/var/lib` is given).

NotAShelf replied:
I do not like the idea of restricting the user to `/var/lib` and allow an override only by overriding the systemd service. In my case, for example, I have a detachable storage mount that I want all apps to dump their state in.

I think using `WorkingDirectory` with an absolute path for is the approach here.

ambroisie commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,148 @@
+    users.users = mkIf (cfg.user == "codeberg-pages") {
+      codeberg-pages = {
+        home = cfg.stateDir;
+        useDefaultShell = true;
```
Why?

NotAShelf commented:
Struggling to add tests here, the Pages server doesn't appear to like being ran inside a sandbox...

teutat3s commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,172 @@
+        default = { };
+        example = {
+          ACME_ACCEPT_TERMS = true;
+          ENABLE_HTTP_SERVER = true;
```
```suggestion
          ACME_ACCEPT_TERMS = "true";
          ENABLE_HTTP_SERVER = "true";
```

teutat3s commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,172 @@
+          HOST = "[::]";
+          PORT = 443;
+
+          GITEA_ROOT = "git.exampledomain.tld";
```
```suggestion
          GITEA_ROOT = "https://git.exampledomain.tld";
```

teutat3s commented on nixos/modules/services/web-apps/codeberg-pages.nix:
```diff
@@ -0,0 +1,172 @@
+          ENABLE_HTTP_SERVER = true;
+
+          HOST = "[::]";
+          PORT = 443;
```
```suggestion
          PORT = "443";
```

teutat3s reviewed (commented):
Tested working :+1: with the above small type changes.

NotAShelf commented:
Thank you for the review @teutat3s, I've addressed your comments.

That should be enough to get codeberg pages going, but I'm (still) a little stuck adding the tests; codeberg pages refuses to connect to a forgejo instance inside my test setup, and the errors are quite crypting. We *could* try to get this merged without tests, but I don't see that happening. Nor is it a good idea.

NotAShelf commented:
Oh, I've been working on them with a standalone repo. I'll push them when I'm on my system.

NotAShelf commented:
Ouch...
